### ISSUE TITLE
Question about the whole React + ProseMirror integration

### ISSUE THREAD

**Issue by TeemuKoivisto** [ifiokjr/remirror#95]
Hey there, been doing rich-text editing for a while and got stuck with a performance problem with Slate so I decided to see if ProseMirror would be the answer.

Since I am, I guess, an idiot, I am hell-bent on doing this app with React so I can't leverage something like TipTap, which has led me to write this short story on this repo. From what I have gathered glueing React and PM together isn't a walk in a park as you might think it is, and there's definitely a high learning-curve for knowing how to implement that *in addition* to PM's quite verbose API.

Anyway, I'm glad somebody's up for the challenge since my brains are still being overloaded with having to go through the rabbit hole of rich text editing and the complex API it forces most libraries to use. So out of curiosity and because I really want to do this, how are you accomplishing this feat of engineering?

My current knowledge about the issue:

* In rich text editing where you edit DOM elements directly, you must re-implement many of the native rich-text functionality yourself since the API quite frankly sucks and is not consistent across different browsers
* All rich-text editing libraries do this at some-level, which requires intricate DOM manipulations with ranges and selections and whatnot, a lot of stuff
* This makes the using of frameworks such as React or Vue less applicable, as you can't let them handle the DOM for you
* However, the creation of complex DOM elements for rich text is such a pain in the ass, using some framework would be preferable. Nobody likes to work directly with DOM and frameworks offer much wanted safety
* So in order to use framework, such as React, you must externalize the DOM element representation to it while at the same time maintaining the rich-text editing semantics of selections and ranges and so on
* ProseMirror offers quite robust way of doing rich-text editing and a way of externalizing that DOM representation through things called NodeViews
* To wrap a NodeView as a React component you must however first establish its ProseMirror schema eg.
```tsx
this.editorState = EditorState.create({
  schema: new Schema({
    nodes: {
      doc: {
        content: 'block+'
      },
      underline: {
        group: 'block',
        content: 'inline*',
        parseDOM: [{ tag: 'p' }],
        toDOM() { return ['p', 0] }
      },
      text: {
        group: 'inline'
      },
    }
  })
})
```
* Which is quite a new frontier for many developers not familiar with PM & rich-text
* Only after that you can separate the NodeView, the rendering itself, by eg (full source copied/modified from here https://gist.github.com/esmevane/7326b19e20a5670954b51ea8618d096d)
```tsx
class Underlined extends React.Component<any, any> {
  public hole: React.RefObject<HTMLParagraphElement>

  constructor(props: any) {
    super(props)
    this.hole = React.createRef()
  }

  public append(node: HTMLElement) {
    if (this.hole) {
      this.hole.current!.appendChild(node)
    }
  }

  public render() {
    return <p ref={this.hole} style={{textDecoration: 'underline'}}></p>
  }
}

class Underline {
  public dom: HTMLElement
  public contentDOM: HTMLElement
  public ref: React.RefObject<any>

  constructor(public node: Node) {
    this.ref = React.createRef()
    this.dom = document.createElement('span')
    this.contentDOM = document.createElement('span')

    ReactDOM.render(
      <Underlined ref={this.ref} />,
      this.dom,
      this.putContentDomInRef
    )
  }

  public update(node: Node) {
    return true
  }

  private putContentDomInRef = () => {
    this.ref.current.append(this.contentDOM)
  }

  destroy() {
    ReactDOM.unmountComponentAtNode(this.dom)
  }
}
```
* Which is the most minimal setup I can think of to integrate PM with React. Still it is a far away from being finished as you can see from my comments of the Gist.

Now that I have laid down my current know-how about the subject, if you have the time could you fill in some blanks for me. Would be highly appreciated.

* I am curious what are the implications of using React instead of direct DOM manipulation that I should know?
* I probably should use ReactDOM's portals as they allow sharing of React context between separately mounted DOM nodes, but is that only the reason why you should use them instead of numerous silly `ReactDOM.render` -calls?
* Then what about the state-management, does is basically happen for all custom events in PM through these `tr.setMeta`/`tr.getMeta` -calls?
* I'm not completely sure what all the Atlassian's editor's boilerplate are for (eg https://bitbucket.org/atlassian/atlaskit-mk-2/src/0fcae893b790443a30f7dadae00638d6e4238b2f/packages/editor/editor-core/src/event-dispatcher/index.ts?at=master)
* As I mentioned in the comments of that Gist it would be nice to have a working minimal PM + React editor with all the required boilerplate. Do you have any idea how to accomplish that?

Hmm.. I guess that's most of it, for now : ). So there's a lot of confusing stuff here that's throwing me off. I really like the idea of establishing a familiar interface to interact with the DOM, like in Slate, but which would actually be performant (and have better API and no ImmutableJS). You have accomplished a lot with your project but it seems it's still far off from being actually usable hence I'd rather just do the PM+React integration myself.

And sorry for the ramble but I've been stuck with this problem for quite a while and I am getting frustrated with the massive complexity of this problem. Martin has spent years on this stuff which I feel I should have too to understand it, or at least taken a course or having read a book about it.

Teemu

**Comment by ifiokjr** [comment]
@TeemuKoivisto thanks for the questions and patience. I fell a bit ill last week so wasn't able to respond as soon I would have like to.

<br>

> I am curious what are the implications of using React instead of direct DOM manipulation that I should know?

I actually think DOM manipulation is better for simpler prosemirror functionality but I can see where the React integration will be helpful for more complex components. The main reason for this is that the current abstraction that I've copied from `atlaskit` isn't yet intuitive to build upon. It still relies on holding a deep understanding of Prosemirror.

<br>

> I probably should use ReactDOM's portals as they allow sharing of React context between separately mounted DOM nodes, but is that only the reason why you should use them instead of numerous silly ReactDOM.render -calls?

Truth be told initially I tried with `ReactDOM.render` since it was easier to start off with (based on the very gist you linked to). For some reason, probably my own fault, it didn't remove the element `onDestroy()`. The node view would be deleted but the react debugger would keep hold of the shadow react element (memory leak). `React.Portal` just worked so I stuck with it. In hindsight, it seems like the better choice, since it allows React to handle the dom stuff.

<br>
> As I mentioned in the comments of that Gist it would be nice to have a working minimal PM + React editor with all the required boilerplate. Do you have any idea how to accomplish that?

The closest I've seen to that is available in [`react-prosemirror`](https://github.com/hubgit/react-prosemirror). 

> I'm not completely sure what all the Atlassian's editor's boilerplate are for (eg https://bitbucket.org/atlassian/atlaskit-mk-2/src/0fcae893b790443a30f7dadae00638d6e4238b2f/packages/editor/editor-core/src/event-dispatcher/index.ts?at=master)

Neither am I, good sir. Neither am I.

**Comment by TeemuKoivisto** [comment]
Hey, thanks for answering. Sorry for the delay, I originally had written a reply soon after you answered but since I felt a little inadequate about my knowledge about this stuff, I decided to work on my own editor to give a more descriptive reply.

I ended up with creating a secondary repo to write all the boilerplate as I figure stuff out, to which I'm soon going to add my Toolbar component that I just created https://github.com/TeemuKoivisto/prosemirror-react-typescript-example

Yeah I spent some time looking into the Atlaskit editor-core, which I managed to run locally all right  (with some problems, as its dependencies are a little bit out of sync..) which has quite a bit of moving parts. Yet I noticed it has its own issues with the context-wrappers sometimes nesting the components *very* deep, like 20-level nesting for a table component. I think that's ridiculous and silly, so something better should be invented for that. *Actually now that I've had a deeper look into it, I must say its code-base is quite horrible. There must have been a good idea in the beginning, but it has bloated out of control.

But there's some really good stuff there like the Table-component I mentioned and I'll be looking into the PortalProvider logic & plugin eventDispatcher if I use them in my work. I'm not interested in creating a creating my own editor library, just getting the basic boilerplate right. I love simple systems which we hold to high esteem at my work, and whatever I'm going to do I'll try to do as simply as possible.

After your reply I spent some time on creating a basic Toolbar with buttons for toggling marks and turning active whenever cursor or selection is on marked text. It helped me to clarify how the state-management should be done, since triggering updates from any editor-transactions to the React components is a bit tricky business. I decided to lean on to the pre-built plugin-state logic, so that all editor-specific state will be there. But instead of hacking some event dispatcher to notify React components about the changes, I'll leverage Mobx observables inside the plugin-state and provide the state slice as probably injected state (not passed down as props). If you're familiar with Mobx you know what I'm aiming at.

I think observables are the simplest way of doing state-management with React and omit a lot of the Redux boilerplate that I've come to hate. My biggest concern is can the UI somehow go out of sync with the PM editor. I mean it shouldn't because React receives the state from PM, which updates everything in singular transactions. Well only if there's state in both React and PM that do something that overlap. Will see how it goes.

If you have any tips or help, I'd appreciate it. =)

**Comment by ifiokjr** [comment]
> I ended up with creating a secondary repo to write all the boilerplate as I figure stuff out, to which I'm soon going to add my Toolbar component that I just created [TeemuKoivisto/prosemirror-react-typescript-example]

Sounds exciting @TeemuKoivisto! I hope this project can be of help in the future.
